test/distributed/fsdp/test_fsdp_hybrid_shard.py

```diff
@@ -0,0 +1,245 @@
+# Owner(s): ["oncall: distributed"]
+
+import contextlib
+import functools
+from functools import partial
+from collections import Counter
+import sys
+
+import torch
+import torch.nn as nn
+import torch.distributed as dist
+
+from torch.distributed.distributed_c10d import _rank_not_in_group
+from torch.distributed.fsdp import (
+    FullyShardedDataParallel as FSDP,
+    ShardingStrategy,
+)
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy, ModuleWrapPolicy
+from torch.nn import TransformerDecoderLayer, TransformerEncoderLayer
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import (
+    CUDAInitMode,
+    FSDPInitMode,
+    FSDPTest,
+    TransformerWithSharedParams,
+)
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    run_tests,
+    TEST_WITH_DEV_DBG_ASAN,
+)
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+@contextlib.contextmanager
+def patch_allreduce(new_allreduce):
+    """
+    Patches dist.all_reduce with a new all_reduce and
+    restores upon exiting.
+    """
+    orig_ar = dist.all_reduce
+    dist.all_reduce = new_allreduce
+    try:
+        yield
+    finally:
+        dist.all_reduce = orig_ar
+
+@contextlib.contextmanager
+def patch_reduce_scatter(new_reduce_scatter):
+    """
+    Patches dist.reduce_scatter_tensor with a new reduce_scatter_tensor and
+    restores upon exiting.
+    """
+    orig_reduce_scatter = dist.reduce_scatter_tensor
+    dist.reduce_scatter_tensor = new_reduce_scatter
+    try:
+        yield
+    finally:
+        dist.reduce_scatter_tensor = orig_reduce_scatter
+
+
+class MyModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.lin1 = nn.Linear(10, 10)
+        self.lin2 = nn.Linear(10, 10)
+        self.lin3 = nn.Linear(10, 10)
+
+class TestFSDPHybridShard(FSDPTest):
+    @property
+    def world_size(self):
+        return max(torch.cuda.device_count(), 2)
```

Comment:
Curious: What is the purpose of this `max()` if we skip if less than 2 GPUs?

Reply:
This was mostly to make CI happy, when there are no cuda devices, this would return 0, which would cause issues in test setup before the tests are skipped.

Reply:
Do you know why we do not need to do this for any of the other FSDP test files? (E.g. is it due to some process group initialization specific to this file?)

Reply:
This was mostly to fix a CI issue, when there are no GPUs but this code still runs.